server/src/main/java/org/elasticsearch/lucene/queries/TimestampTwoPhaseIterator.java

```diff
@@ -0,0 +1,207 @@
+                        nextLevel++;
+                    }
+                }
+                switch (match) {
+                    case YES:
+                        return doc = target;
+                    case MAYBE:
+                        if (target > innerApproximation.docID()) {
+                            target = innerApproximation.advance(target);
+                        }
+                        if (target <= upTo) {
+                            return doc = target;
+                        }
+                        break;
+                    case NO_AND_SKIP:
+                        if (target > primaryFieldUpTo) {
+                            primaryFieldSkipper.advance(target);
+                            for (int level = 0; level < primaryFieldSkipper.numLevels(); level++) {
+                                if (primaryFieldSkipper.minValue(level) == primaryFieldSkipper.maxValue(level)) {
+                                    primaryFieldUpTo = primaryFieldSkipper.maxDocID(level);
+                                } else {
+                                    break;
+                                }
+                            }
+                            if (primaryFieldUpTo > upTo) {
+                                upTo = primaryFieldUpTo;
+                            }
+                        }
+                        if (upTo == DocIdSetIterator.NO_MORE_DOCS) {
+                            return doc = NO_MORE_DOCS;
+                        }
+                        target = upTo + 1;
+                        break;
+                    case NO:
+                        if (upTo == DocIdSetIterator.NO_MORE_DOCS) {
+                            return doc = NO_MORE_DOCS;
+                        }
+                        target = upTo + 1;
+                        break;
+                    default:
+                        throw new AssertionError("Unknown enum constant: " + match);
+                }
+            }
+        }
+
+        @Override
+        public long cost() {
+            return innerApproximation.cost();
+        }
+
+        Match match(int level) {
+            long minValue = timestampSkipper.minValue(level);
+            long maxValue = timestampSkipper.maxValue(level);
+            if (minValue > maxTimestamp) {
+                return Match.NO;
+            } else if (maxValue < minTimestamp) {
+                return Match.NO_AND_SKIP;
+            } else if (minValue >= minTimestamp && maxValue <= maxTimestamp) {
+                return Match.YES;
+            } else {
+                return Match.MAYBE;
+            }
+        }
+
+    }
+
+    @Override
+    public boolean matches() throws IOException {
+        return switch (approximation.match) {
+            case YES -> true;
+            case MAYBE -> {
+                final long value = timestamps.longValue();
+                yield value >= minTimestamp && value <= maxTimestamp;
```

Comment:
if value < minTimestamp, we know we will not be matching any more documents, maybe we can move approximation.match to NO?

Reply:
Interesting, let me take a look at this.

Reply:
done: 57395650b39aa637db395b059c1740b8cdf82bda